File: mill-example/oystercard/src/oystercard/java/OysterCard.java
package oystercard.java;

import java.util.Optional;

/**
 * The main entry point which represents an OysterCard's state.
 *
 *
 * It's a persistent data structure which can be updated via 'update' or explicitly calling 'takeBus', 'loadCard', or 'touchCard',
 * all of which return a new OysterCard.
 *
 * If used within a context which actually will make a charge to some database/disk, one could just take a difference of a
 * before and after balance.
 *
 */
public final class OysterCard {
    private final Optional<TouchIn> lastTouch;
    public final Amount balance;

    public static OysterCard Empty = new OysterCard();

    public OysterCard() {
        this(Amount.inPence(0));
    }
    public OysterCard(Amount startingBalance) {
        this(startingBalance, Optional.empty());
    }
    private OysterCard(Amount startingBalance, Optional<TouchIn> lastTouch) {
        this.lastTouch = lastTouch;
        this.balance = startingBalance;
    }
    public OysterCard update(UserEvent event, JourneyCalculator calculator) {
        if (event.equals(UserEvent.takeBus)) {
            return takeBus(calculator.busFare());
        } else if (event instanceof UserEvent.LoadCard) {
            return loadCard(((UserEvent.LoadCard) event).amount);
        } else if (event instanceof UserEvent.TouchCard) {
            return touchCard(((UserEvent.TouchCard) event).station, calculator);
        } else {
            throw new IllegalArgumentException("Unknown user event " + event);
        }
    }

    @Override
    public String toString() {
        return "OysterCard(" + balance + ", " + lastTouch +")";
    }

    public OysterCard takeBus(Amount busFare) {
        return new OysterCard(balance.minus(busFare));
    }
    public OysterCard loadCard(Amount amount) {
        return new OysterCard(balance.plus(amount));
    }
    public OysterCard touchCard(TubeStation station, JourneyCalculator calculator) {
        return lastTouch.map(touchIn -> touchOut(touchIn, station, calculator)).orElse(touchIn(station, calculator));
    }

    private OysterCard touchIn(TubeStation station, JourneyCalculator calculator) {
        Amount fare = calculator.maxFare();
        return new OysterCard(balance.minus(fare), Optional.of(new TouchIn(station, fare)));
    }

    private OysterCard touchOut(TouchIn touchedIn, TubeStation touchOut, JourneyCalculator calculator) {
        Amount cost = calculator.price(touchedIn.station, touchOut);
        Amount newBalance = balance.plus(touchedIn.fareCharged).minus(cost);
        return new OysterCard(newBalance);
    }
}

File: mill-example/oystercard/src/oystercard/scala/OysterCard.scala
package oystercard.scala

/**
  * A persistent (immutable) data structure representing a user's oyster card.
  *
  * An OysterCard itself is immutable, and can be updating using the DSL-like functions [[touchIn]], [[topUpCard(Int)]] or [[takeTheBus]]
  * if it is being accessed programatically:
  *
  * {{{
  *   val initial : OysterCard = ...
  *   val updatedCard : OysterCard = initial.topUpCard(2000).touchCard(Hammersmith)
  *
  *   // the 'initial' and 'updatedCard' balances will differ as 'initial' is unchanged
  * }}}
  *
  * It can also be updated via the [[update]] function which takes a [[UserEvent]]s.
  *
  * For example, if we had a stream (or observable) of user inputs, perhaps from a message queue or REST endpoint:
  *
  * {{{
  *   val userInputs : Observable[UserEvent] = ... // some user inputs
  *   implicit val calculator : JourneyCalculator = JourneyCalculator()
  *   val updatedBalances : Observable[Amount] = userInputs.foldLeft(OysterCard()) {
  *     case (card, userEvent) => card.update(userEvent)
  *   }.map(_.balance)
  * }}}
  *
  * All actions other than 'topUpCard' require a [[JourneyCalculator]] parameter which is used to provide the fare prices for journeys.
  *
  *
  * @param balance   the current balance
  * @param lastTouch the last tube station touched in if one exists
  */
final case class OysterCard(balance: Amount, lastTouch: Option[TouchIn] = None) {

  /** This function will return a new OysterCard representing the updated state
    *
    * @param event the user event -- this could have been unmarshalled from some POST request or from some input stream
    * @param calculator a calculator used to price journeys
    * @return an updated oyster card
    */
  def update(event: UserEvent)(implicit calculator: JourneyCalculator): OysterCard = {
    event match {
      case TouchCard(station) => touchCard(station)
      case TakeBus => takeTheBus
      case LoadCard(amount) => topUpCard(amount)
    }
  }

  def topUpCard(amount: Amount) = copy(balance = balance + amount)

  def touchCard(station: TubeStation)(implicit calculator: JourneyCalculator): OysterCard = {
    lastTouch match {
      case None => touchIn(station)
      case Some(lastTouch) => touchOut(lastTouch, station)
    }
  }

  def takeTheBus(implicit calculator: JourneyCalculator): OysterCard = copy(balance = balance - calculator.busFare)

  // useful for tests - updates a card w/ all the given events
  def updateAll(events: Iterable[UserEvent])(implicit calculator: JourneyCalculator): OysterCard = {
    events.foldLeft(this) {
      case (card, action) => card.update(action)
    }
  }

  private def touchIn(station: TubeStation)(implicit calculator: JourneyCalculator): OysterCard = {
    copy(balance = balance - calculator.maxFare, lastTouch = Option(TouchIn(station, calculator.maxFare)))
  }

  private def touchOut(stationTouchedIn: TouchIn, stationTouchedOut: TubeStation)(implicit calculator: JourneyCalculator): OysterCard = {
    val newBalance = balance + stationTouchedIn.fareCharged - calculator.price(stationTouchedIn.at, stationTouchedOut)
    copy(balance = newBalance, lastTouch = None)
  }

}

object OysterCard {
  def apply() = new OysterCard(Amount.Zero, None)
}